### PR TITLE
Fix regression test 46239 with Crossgen2 and improve runtime logging

### PR DESCRIPTION
**Pull request by trylek**
The regression test

<code>src\tests\JIT\Regressions\JitBlue\Runtime_46239</code>

exercises various interesting corner cases of type layout that
weren't handled properly in Crossgen2 on x86 and ARM[32]. This
change fixes the remaining deficiencies and it also adds
provisions for better runtime logging upon type layout mismatches.

Thanks

Tomas

/cc @dotnet/crossgen-contrib 

P.S. For now I don't expect we need to push this to Preview 4 bits
as the change covers a rather corner case on the less prevalent
architectures; please let me know if you believe otherwise.

### CODE REVIEW

File: src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs [hunk 1]
```diff
@@ -304,17 +305,19 @@ protected virtual void FinalizeRuntimeSpecificStaticFieldLayout(TypeSystemContex
         {
         }
 
-        protected static ComputedInstanceFieldLayout ComputeExplicitFieldLayout(MetadataType type, int numInstanceFields)
+        protected virtual bool IsBlittableOrManagedSequential(TypeDesc type) => false;
```
Comment (davidwrighton) [inline src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs:308]: I suspect the default here should be true. Blittable types have a requirement that they are handled in the same way on all runtimes as they describe native data structures, and by defaulting to false, this leads for a way for the algorithm to misbehave on Native AOT.

I'd prefer to see this either be default true, or make it an abstract method and force the NativeAOT compiler to deal with this when it gets there.

**Review by davidwrighton — CHANGES_REQUESTED** [review]
Other than my comment on the IsBlittableOrManagedSequential, I'm pretty happy with this.

File: src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs (hunk 1 above)
Comment (davidwrighton) [reply src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs:308]: @MichalStrehovsky could you comment here on what you'd like to see.

Comment (MichalStrehovsky) [reply src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs:308]: Maybe let's just move `IsManagedSequentialType` to here and then this doesn't even need to be virtual and can just do the same thing?

We didn't have trouble with this in .NET Native (where this is missing) so I don't think it matters, but we ported over so many CoreCLR weirdness's to this algorithm over time that I no longer understand how layout is done and would prefer all bugs in it to be also surfaced in crossgen2 (and not have NativeAOT specific bugs that we need to troubleshoot there).

**Comment by BruceForstall** [comment]
It looks like this is causing every jitstress job flavor to fail:

https://dev.azure.com/dnceng/public/_build/results?buildId=1097089&view=ms.vss-test-web.build-test-results-tab&runId=33568136&resultId=111051&paneView=debug

**Comment by BruceForstall** [comment]
Any chance you'll get this in soon? there are a lot of test jobs failing due to test49826.

File: src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs (hunk 1 above)
Comment (trylek) [reply src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs:308]: @MichalStrehovsky - Thanks for your feedback. I'm trying to modify the change based on your suggestion but I'm hitting a layering problem - IsBlittable is part of MarshalUtils and that's not in ILCompiler.TypeSystem.ReadyToRun and it seems to me that moving it over brings various bits of interop code into ILCompiler.TypeSystem.ReadyToRun which I suspect to be undesirable as my current understanding is that NativeAOT doesn't use exactly identical interop as Crossgen2 / CoreCLR. What do you think would be the ideal way to resolve this discrepancy?

Comment (MichalStrehovsky) [reply src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs:308]: MarshalUtils are part of ILCompiler.TypeSystem on the NativeAOT side, so I'm not opposed to moving it (and things it depends on) into ILCompiler.TypeSystem.ReadyToRun as well. The split is rather arbitrary and interop is getting tangled into the type system the same way as is on CoreCLR. It was a nice dream to have a separation of concerns there.

https://github.com/dotnet/runtimelab/blob/c8e3158b0981419527b18f3c0259bcf02103e054/src/coreclr/tools/aot/ILCompiler.TypeSystem/ILCompiler.TypeSystem.csproj#L528-L530

Comment (trylek) [reply src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs]: Thanks Michal for your feedback. I have updated the change based on your suggestion, please take another look when you have a chance and let me know if you find the project shuffle acceptable.

Comment (jkotas) [reply src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs]: >  The split is rather arbitrary and interop is getting tangled into the type system the same way as is on CoreCLR. It was a nice dream to have a separation of concerns there.

Would there be a way to adjust the runtime side of the things to avoid this mess?

**Review by davidwrighton — DISMISSED** [review]

File: src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs (hunk 1 above)
Comment (trylek) [reply src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs]: Well, for now I think I have at least managed to make something like an inventory of the various runtime behaviors by implementing them in Crossgen2. In theory we may be able to simplify some of the stuff, I however suspect that some of the subtle distinctions amount to tiny memory use optimizations on the runtime side i.e. something that's very tricky to undo as it may incur working set regressions in arbitrary scenarios.

Comment (jkotas) [reply src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs]: > however suspect that some of the subtle distinctions amount to tiny memory use optimizations on the runtime side

I doubt that it would show up on the radar. The behavior of the field layout algorithm is completely accidental in number of cases, and in fact there are many known situations where it is inefficient for no good reason.

Comment (jkotas) [reply src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs]: Examples:
- https://github.com/dotnet/coreclr/pull/27195#discussion_r334985489
- https://github.com/dotnet/runtime/issues/12977

Comment (trylek) [reply src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs]: You're probably the biggest expert in this field so I have no desire to doubt your assessment. My only point is that further independent development of runtime vs. Crossgen2 side of this equation is giving me goosebumps; my current expectation is that we'll ultimately go down the path you yourself suggested in the past - having a way to communicate the field layouts from the compiler to the runtime, that will give us room for experimentation with potential packing optimizations and rid us of the burdensome need to keep both codebases in 100% sync.

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs [hunk 2]
```diff
@@ -2257,7 +2249,7 @@ private bool NeedsTypeLayoutCheck(TypeDesc type)
 
         private bool HasLayoutMetadata(TypeDesc type)
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:2250]: The one place where this method is used looks like dead/unrechable code. 

This method can only return true when `type.IsValueType` is true, but this case is handled in `EncodeFieldBaseOffset` earlier via the `else if (pMT.IsValueType)` check. It means that this method won't ever return true.

Comment (trylek) [reply src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs]: Nice cleanup, thanks Jan for pointing that out, deleted in 9th commit.

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs [hunk 3]
```diff
@@ -2327,17 +2308,6 @@ private void EncodeFieldBaseOffset(FieldDesc field, CORINFO_FIELD_INFO* pResult,
                 }
                 // ENCODE_NONE
             }
-            else if (HasLayoutMetadata(pMT))
-            {
-                PreventRecursiveFieldInlinesOutsideVersionBubble(field, callerMethod);
-
-                // We won't try to be smart for classes with layout.
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:2334]: Actually, the old crossgen has a special case for this with this comment that kicks in for non-value types (different from what was here before that kicked in for value types).

Do we want to match it? Or are we confident that we have extensive test coverage for all non-value types with layout corner cases?

Also, once we add this back, can at least some of the remaining changes be simplified?

Comment (trylek) [reply src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:2334]: According to my understanding of the CoreCLR code, the layout check is based on the following condition:

https://github.com/dotnet/runtime/blob/207b03a073107e63df0b67056dd596226065651e/src/coreclr/vm/methodtablebuilder.cpp#L11924

It seems to me that this means that the conditional statement you originally pointed out should basically stay, the problematic bit is the method HasLayoutMetadata itself - we're certainly trying to match the CoreCLR runtime behavior, in this particular case we probably just missed the fact that the CoreRT version of HasLayoutMetadata is quite different than its CoreCLR runtime counterpart method

https://github.com/dotnet/runtime/blob/207b03a073107e63df0b67056dd596226065651e/src/coreclr/vm/methodtablebuilder.cpp#L11746

I'll look into putting HasLayoutMetadata back and modifying it based on the CoreCLR runtime version. Other than that, can you please be more specific regarding the simplifications you're suggesting? While I do agree that some of the runtime logic is complicated and somewhat arbitrary, I'm worried that unsettling it at this point might cause more harm than good.

Comment (jkotas) [reply src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:2334]: I think the equivalent of the HasLayoutMetadata should be just `type.IsSequentialLayout || type.IsExplicitLayout`.

Yes, the runtime logic is complicated. We have done multiple rounds of trying to replicate it, and we are still finding mismatches. It is safe bet that this is not the last round of fixes. I think we need to make it simpler to get confident that it matches. I will do some ad-hoc testing to try to find more situations where it does not match.

Comment (trylek) [reply src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs]: Thanks Jan for your additional feedback. I have updated the runtime check based on your suggestion. Once the change passes basic testing, I'll also trigger another round of Pri1 testing that previously uncovered most of these inconsistencies. It would be certainly awesome if you were willing to use your vast expertise in this area to devise additional tests exercising various obscure corner cases we may have previously missed.

File: src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs [hunk 4]
```diff
@@ -857,7 +915,9 @@ private static SizeAndAlignment ComputeInstanceSize(MetadataType type, LayoutInt
             {
                 if (type.IsValueType)
                 {
-                    instanceSize = LayoutInt.AlignUp(instanceSize, alignment, target);
+                    instanceSize = LayoutInt.AlignUp(instanceSize,
+                        alignUpInstanceByteSize ? alignment : LayoutInt.Min(alignment, target.LayoutPointerSize),
```
Comment (jkotas) [inline src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs:919]: I assume that the `alignUpInstanceByteSize = false` case will kick in for a struct like this on ARM32: 
```
    [StructLayout(LayoutKind.Explicit)]
    internal struct S3
    {
        [FieldOffset(0)]
        public ulong tmp1;
        [FieldOffset(8)]
        public Object tmp2;
    }
```
The algorithm will compute the instanceSize of this struct as 12. Is that correct?

It is a bug in the type loader. The instance size of this struct on ARM should be 16, so that the long field is aligned at 8 bytes, so that potential atomic 64-bit operations work fine on it. It would be better to fix the type loader instead of replicating the bug here.

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs [hunk 5]
```diff
@@ -2257,13 +2249,7 @@ private bool NeedsTypeLayoutCheck(TypeDesc type)
...
         private bool HasLayoutMetadata(TypeDesc type)
         {
-            if (type.IsValueType && (MarshalUtils.IsBlittableType(type) || ReadyToRunMetadataFieldLayoutAlgorithm.IsManagedSequentialType(type)))
-            {
-                // Sequential layout
-                return true;
-            }
-
-            return false;
+            return MetadataFieldLayoutAlgorithm.IsBlittableOrManagedSequentialType(type) || (type is MetadataType mt && mt.IsExplicitLayout);
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:2252]: I would simplify this to `mt.IsExplicitLayout || mt.IsSequentialLayout`. Maybe even inline it at the callsite.

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunMetadataFieldLayoutAlgorithm.cs [hunk 6]
```diff
@@ -803,7 +803,7 @@ protected override ComputedInstanceFieldLayout ComputeInstanceFieldLayout(Metada
                 return ComputeExplicitFieldLayout(type, numInstanceFields);
             }
             else
-            if (type.IsEnum || MarshalUtils.IsBlittableType(type) || IsManagedSequentialType(type))
+            if (type.IsEnum || IsBlittableOrManagedSequentialType(type))
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunMetadataFieldLayoutAlgorithm.cs:806]: To get rid of dependency of the type layout algorithm on blittability here, I think we would simplify this into:

`if (type.IsSequential && type.ContainsPointers)`

(with matching change in the runtime). It would be R2R breaking change, but it would set us up for future. 

The dependency of the layout algorithm on blittability makes changes like #103 R2R breaking changes. We did not treat #103 as a R2R breaking change since the connection between interop, type layout and R2R was not understood.

File: src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs [hunk 7]
```diff
@@ -304,17 +304,17 @@ protected virtual void FinalizeRuntimeSpecificStaticFieldLayout(TypeSystemContex
         {
         }
 
-        protected static ComputedInstanceFieldLayout ComputeExplicitFieldLayout(MetadataType type, int numInstanceFields)
+        protected ComputedInstanceFieldLayout ComputeExplicitFieldLayout(MetadataType type, int numInstanceFields)
```
Comment (davidwrighton) [inline src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs:307]: Is there a reason why you changed this from static to non-static?

Comment (trylek) [reply src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs:307]: At one point it was needed because there used to be a helper method calculating the offset bias. I have double checked it's not longer necessary after all the refactorings, will remove, thanks for pointing that out!

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs [hunk 8]
```diff
@@ -2262,13 +2262,7 @@ private bool NeedsTypeLayoutCheck(TypeDesc type)
...
         private bool HasLayoutMetadata(TypeDesc type)
         {
-            if (type.IsValueType && (MarshalUtils.IsBlittableType(type) || ReadyToRunMetadataFieldLayoutAlgorithm.IsManagedSequentialType(type)))
-            {
-                // Sequential layout
-                return true;
-            }
-
-            return false;
+            return type is MetadataType metadataType && (metadataType.IsSequentialLayout || metadataType.IsExplicitLayout);
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:2265]: Can we also delete/replace `IsBlittableType` from `ComputeInstanceFieldLayout` so that the layout does not depend on the algorithm to compute blittability?

**Comment by trylek** [comment]
@jkotas - I believe I have managed to make some progress on the change based on your suggestions but in accordance with @davidwrighton I think it's even riskier to take than the original attempt using our existing blittability / managed sequential checks. As you can easily imagine, simplifying the condition for using sequential vs. auto layout silently switches over some types from one to the other and the bug tail is enormous, I've been digging through it over the last two weeks and I'm nowhere near green state. Two particular simple examples:

1) GCMemoryInfoData - this managed class has a native runtime counterpart and we use the DEFINE_CLASS / DEFINE_FIELD macros to check their consistency. Without the change the class uses auto layout as it's disqualified from managed sequential due to being a reference type; interestingly enough, once my change switched it over to start using the sequential layout, it fell out of sync on x86 because the sequential layout algorithm is a worse match for the native layout than the auto layout. I have fix for this particular issue but the reality is that the fix technically changes the semantics of sequential layout and potentially affects blittable types and interop in general.

2) LAHashDependentHashTracker - this managed class has a native runtime counterpart just like GCMemoryInfoData but without my change it's considered blittable i.e. eligible for using sequential layout. When I attempted to limit changes to pre-existing behavior by only using sequential layout for value types without GC pointers, it switched this class to auto layout which ended up reordering its two fields and falling out of sync with the runtime.

My current plan is to mark this change as NO-MERGE for now and revive its previous version after incorporating Michal's feedback in form of a new PR; I can return to this cleanup after we fork off the final .NET 6 shipping bits - the 1-2 month period of inability to make more substantial runtime changes should be ideal for repeated testing and polishing aspects of this change. In many cases it will be useful / needed to consult official native compiler documentation to make sure we're doing the right / expected thing.

**Comment by trylek** [comment]
Closing for now; I'll revive the PR after we fork off for .NET 6 and I'll start the new CI testing rounds to iron out the various problems.
